new/prusti-interface/src/specs/typed.rs

```diff
@@ -42,33 +45,77 @@ impl StructuralToTyped<Expression> for json::Expression {
     }
 }
 
-impl StructuralToTyped<AssertionKind> for json::AssertionKind {
-    fn to_typed(self, typed_expressions: &HashMap<String, rustc_hir::BodyId>) -> AssertionKind {
+impl<'tcx> StructuralToTyped<'tcx, TriggerSet> for json::TriggerSet {
+    fn to_typed(self, typed_expressions: &HashMap<String, Node>, tcx: TyCtxt<'tcx>) -> TriggerSet {
+        common::TriggerSet(
+            self.0
+                .into_iter()
+                .map(|x| x.to_typed(typed_expressions, tcx))
+                .collect()
+        )
+    }
+}
+
+impl<'tcx> StructuralToTyped<'tcx, Trigger> for json::Trigger {
+    fn to_typed(self, typed_expressions: &HashMap<String, Node>, tcx: TyCtxt<'tcx>) -> Trigger {
+        common::Trigger(
+            self.0
+                .into_iter()
+                .map(|x| x.to_typed(typed_expressions, tcx))
+                .collect()
+        )
+    }
+}
+
+impl<'tcx> StructuralToTyped<'tcx, ForAllVars<'tcx>> for json::ForAllVars {
+    fn to_typed(self, typed_expressions: &HashMap<String, Node>, tcx: TyCtxt<'tcx>) -> ForAllVars<'tcx> {
+        let hir_id = typed_expressions[&format!("{}_{}", self.spec_id, self.expr_id)].hir_id;
+        let local_id = tcx.hir().local_def_id(hir_id);
+        let (body, _) = tcx.mir_validated(local_id);
+        let body = body.borrow();
+
+        let mut vars = vec![];
+        for local_decl in body.local_decls.iter() {
```

Comment:
Shouldn't this be traversal over closure arguments only and not all variables defined in the closure?

Reply:
yeah, but how to get those?

Reply:
I tried with `body.args_iter()`, but that does not work, since even the count of variables obtained from the parser differs from the `arg_count` of the body. So I'd keep this as is, at least for now.

Reply:
Why it differs? It should be the same.

Reply:
If it is a bit more readable, this works:
```
let mut vars = vec![];
        for local_decl in body.local_decls.iter() {
            if local_decl.is_user_variable() {
                vars.push(local_decl.ty.kind.clone())
            }
        }
```
However, [this iterator](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/mir/mod.rs.html#295-300) does not: it sometimes misses a variable.

Reply:
or maybe this is even a bit more compact:
```
        let vars: Vec<TyKind> = body
            .local_decls
            .iter()
            .filter(|x| x.is_user_variable())
            .map(|x| x.ty.kind.clone())
            .collect();
```

Reply:
But this sounds wrong. You need arguments, not just variables.

Reply:
Are there more or fewer arguments than expected?

Reply:
The iterator misses some, i.e. reports 0 instead of 1. And I think that `local_decls` holds them all, doesn't it?

Reply:
`local_decls` contains arguments, variables, and the return value while we need only arguments.

Reply:
Looking at the `args_iter` [implementation](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/mir/mod.rs.html#334-337), it should return the exact number of closure arguments. If it does not, then something is wrong with the encoding.

Reply:
Apparently it is not the case, when the `assert!(self.count == body.arg_count)` (i.e. check the number of variables from the parser == the number of arguments to the body) fails.

Reply:
Basically, my point is that if `args_iter` does not return what we expect, we have to understand why.

Reply:
> Apparently it is not the case, when the `assert!(self.count == body.arg_count)` (i.e. check the number of variables from the parser == the number of arguments to the body) fails.

What is the exact difference?

Reply:
> Basically, my point is that if `args_iter` does not return what we expect, we have to understand why.

I am just trying to investigate this. Will let you in a few moments once I find something out.

Reply:
So APPARENTLY (i.e. not confirmed by looking into the source code) the first argument is the closure itself. This holds:
```rust
assert!(body.arg_count-1 == self.count);
```
and this works:
```rust
let mut vars = vec![];
for arg in body.args_iter() {
    let decl = body.local_decls.get(arg).unwrap();
    if !decl.is_user_variable() {
        continue;
    }
    vars.push(decl.ty.kind.clone());
}
```

Reply:
Looks good. Could you also add an assert that `vars.len() == self.count`?

Reply:
This one is already there, that is the most basic one.